src/ansiblelint/utils.py

```diff
@@ -901,3 +909,80 @@ def nested_items(
+
+    This is a recursive function that calls itself for each nested layer of data.
+    Each iteration yields:
+
+    1. the current item's dictionary key or list index,
+    2. the current item's value, and
+    3. the path to the current item from the outermost data structure.
+
+    For dicts, the yielded (1) key and (2) value are what mydict.items() yields.
+    For lists, the yielded (1) index and (2) value are what enumerate(mylist) yields.
+    The final component, the parent path, is a list of dict keys and list indexes.
+    The parent path can be helpful in providing error messages that indicate
+    precisely which part of a yaml file (or other data strucutre) needs to be fixed.
+
+    For example, given this playbook:
+
+    .. code-block:: yaml
+
+        - name: a play
+          tasks:
+          - name: a task
+            debug:
+              msg: foobar
+
+    Here's the first and last yielded items:
+
+    .. code-block:: python
+
+        >>> playbook=[{"name": "a play", "tasks": [{"name": "a task", "debug": {"msg": "foobar"}}]}]
+        >>> next( nested_items_path( playbook ) )
+        (0, {'name': 'a play', 'tasks': [{'name': 'a task', 'debug': {'msg': 'foobar'}}]}, [])
+        >>> list( nested_items_path( playbook ) )[-1]
+        ('msg', 'foobar', [0, 'tasks', 0, 'debug'])
+
+    Note that, for outermost data structure, the parent path is ``[]`` because
+    you do not need to descend into any nested dicts or lists to find the indicated
+    key and value.
+
+    If a rule were designed to prohimit "foobar" debug messages, it could use the
+    parent path to provide a path to the problematic ``msg``. It might use a jq-style
+    path in its error message: "the error is at ``.[0].tasks[0].debug.msg``".
+    Or if a utility could automatically fix issues, it could use the path to descend
+    to the parent object using something like this:
+
+    .. code-block:: python
+
+        target = data
+        for segment in parent_path:
+            target = target[segment]
+
+    :param data: The nested data (dicts or lists).
+    :param parent_path: Do not use this param. It is used internally to recursively
+                        build the yielded parent path (list of keys, indexes).
+
+    :returns: each iteration yields the key (of the parent dict) or the index (lists)
+    """
+    if parent_path is None:
+        parent_path = []
+    if isinstance(data, dict):
+        convert_to_tuples = data.__class__.items  # dict subclasses can override items
+    elif isinstance(data, list):
+        convert_to_tuples = enumerate
+    else:
+        raise TypeError(
+            f"nested_items_path expects a dict or a list but got {type(data)}"
+        )
+    for key, value in convert_to_tuples(data):
+        yield key, value, parent_path
+        if isinstance(value, (dict, list)):
+            yield from nested_items_path(data=value, parent_path=parent_path + [key])
```

Comment:
mypy didn't like this refactor.
```
src/ansiblelint/utils.py:980: error: Incompatible types in assignment (expression has type "Type[enumerate[Any]]", variable has type "Callable[[Dict[_KT, _VT]], ItemsView[_KT, _VT]]")
src/ansiblelint/utils.py:985: error: Argument 1 has incompatible type "Union[Dict[Any, Any], List[Any]]"; expected "Dict[_KT, _VT]"
src/ansiblelint/utils.py:988: error: List item 0 has incompatible type "_KT"; expected "Union[str, int]"
```

Any suggestions on how to clean this up for mypy? Or should I go back to how it was?

Reply:
OK. I found a `cast()` that works for this.

Reply:
I think, for this, it's possible to solve with custom type definitions but it'd require some extra effort. This is caused by `Any` being too generic.

Reply:
By the way, one of the ways to narrow types is type guards: https://mypy.readthedocs.io/en/latest/type_narrowing.html#user-defined-type-guards.

Reply:
Oh, I see. mypy gets confused by `if isinstance(value, (dict, list)):` — if you were to check for just one type, it'd understand what's going on better, I think. But that's probably not what you want...

Reply:
He fixe it.